File: app/src/main/java/message/smsapp/message/ConversationManager.java
package message.smsapp.message;

import android.Manifest;
import android.app.Activity;
import android.content.ContentResolver;
import android.content.Context;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.widget.ListAdapter;
import android.widget.ListView;

import java.util.ArrayList;

import message.smsapp.R;
import message.smsapp.layouts_manager.MessagesAdapter;

public class ConversationManager {
	private final Context context;
	
	private ArrayList<String> listOfComponents, listOfContents;
	private ArrayList<Integer> listOfIcons;
	private ContactManager contactManager;
	private ArrayList<Conversation> conversations;
	
	public ConversationManager(Context context, ListView list, int REQUEST_PERMISSION_KEY) {
		this.context = context;
		this.listOfComponents = new ArrayList<>();
		this.listOfContents = new ArrayList<>();
		this.listOfIcons = new ArrayList<>();
		this.conversations = new ArrayList<>();
		this.contactManager = new ContactManager(context, REQUEST_PERMISSION_KEY);
		
		if (ContextCompat.checkSelfPermission(context, Manifest.permission.READ_SMS) !=
				PackageManager.PERMISSION_GRANTED) {
			if (ActivityCompat.shouldShowRequestPermissionRationale((Activity) context,
					Manifest.permission.READ_SMS)) {
				
			} else {
				ActivityCompat.requestPermissions((Activity) context,
						new String[]{Manifest.permission.READ_SMS},
						REQUEST_PERMISSION_KEY);
			}
		}
		
		this.refreshSMS();
		
		ListAdapter adapter = new MessagesAdapter(this.context, listOfComponents, listOfContents,
				listOfIcons);
		list.setAdapter(adapter);
	}
	
	private void refreshSMS() {
		this.clearDatas();
		
		ContentResolver contentResolver = this.context.getContentResolver();
		final String[] projection = new String[]{"*"};
		Uri uri = Uri.parse("content://sms/");
		Cursor cursor = contentResolver.query(uri, projection, null, null, null);
		
		int indexBody = cursor.getColumnIndex("body");
		int indexAddress = cursor.getColumnIndex("address");
		if (indexBody < 0 || !cursor.moveToFirst()) return;
		
		do {
			this.addConversation(this.contactManager.getContactNameWithNumber(cursor.getString
					(indexAddress)), getPreviewBody(cursor.getString(indexBody)), R
					.drawable
					.ic_menu_send);
		} while (cursor.moveToNext());
		
		cursor.close();
		
		this.conversations = Conversation.generateConversation(listOfComponents, listOfContents);
		
		this.generateListWithConversations();
	}
	
	private void generateListWithConversations(){
		ArrayList<String> temp = new ArrayList<>();
		
		for (int i = 0; i < this.conversations.size(); i++) {
			temp.add(this.conversations.get(i).getPhoneNumber());
		}
		
		this.listOfComponents = temp;
		
		temp = new ArrayList<>();
		
		for (int i = 0; i < this.conversations.size(); i++) {
			temp.add(this.conversations.get(i).getMessages().get(0));
		}
		
		this.listOfContents = temp;
		
		this.addConversation(Integer.toString(this.conversations.size()), "test", R.drawable.ic_menu_camera);
	}
	
	@NonNull
	private String getPreviewBody(String body) {
		String preview = body.substring(0, Math.min(body.length(), 35));
		preview += (body.length() > preview.length()) ? "..." : "";
		
		return preview;
	}
	
	private void addConversation(String title, String content, int id) {
		this.listOfComponents.add(title);
		this.listOfContents.add(content);
		this.listOfIcons.add(id);
	}
	
	private void clearDatas() {
		this.listOfComponents.clear();
		this.listOfContents.clear();
		this.listOfIcons.clear();
		this.conversations.clear();
	}
}

File: app/src/main/java/message/smsapp/message/ConversationManager.kt
package message.smsapp.message

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.net.Uri
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.widget.ListAdapter
import android.widget.ListView
import message.smsapp.R
import message.smsapp.layouts_manager.MessagesAdapter
import java.util.*

class ConversationManager(private val context: Context, list: ListView, REQUEST_PERMISSION_KEY: Int) {
    private var listOfComponents: ArrayList<String>
    private var listOfContents: ArrayList<String>
    private val listOfIcons: ArrayList<Int>
    private val contactManager: ContactManager
    private var conversations: ArrayList<Conversation>

    private fun refreshSMS() {
        clearDatas()
        context.contentResolver.query(Uri.parse("content://sms/inbox"), arrayOf("*"), null, null, null)?.use {
            val indexBody = it.getColumnIndex("body")
            val indexAddress = it.getColumnIndex("address")
            if (indexBody < 0 || !it.moveToFirst()) return
            do {
                addConversation(contactManager.getContactNameWithNumber(it.getString(indexAddress)), getPreviewBody(it.getString(indexBody)), R.drawable.ic_menu_send)
            } while (it.isAfterLast)
        }
        conversations = Conversation.generateConversation(listOfComponents, listOfContents)
        generateListWithConversations()
    }

    private fun generateListWithConversations() {
        var temp = ArrayList<String>()
        for (i in conversations.indices) {
            temp.add(conversations[i].phoneNumber)
        }
        listOfComponents = temp
        temp = ArrayList()
        for (i in conversations.indices) {
            temp.add(conversations[i].messages[0])
        }
        listOfContents = temp
        addConversation(conversations.size.toString(), "test", R.drawable.ic_menu_camera)
    }

    private fun getPreviewBody(body: String): String {
        var preview = body.substring(0, body.length.coerceAtMost(35))
        preview += if (body.length > preview.length) "..." else ""
        return preview
    }

    private fun addConversation(title: String, content: String, id: Int) {
        listOfComponents.add(title)
        listOfContents.add(content)
        listOfIcons.add(id)
    }

    private fun clearDatas() {
        listOfComponents.clear()
        listOfContents.clear()
        listOfIcons.clear()
        conversations.clear()
    }

    init {
        listOfComponents = ArrayList()
        listOfContents = ArrayList()
        listOfIcons = ArrayList()
        conversations = ArrayList()
        contactManager = ContactManager(context, REQUEST_PERMISSION_KEY)
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.READ_SMS) !=
                PackageManager.PERMISSION_GRANTED) {
            if (ActivityCompat.shouldShowRequestPermissionRationale((context as Activity),
                            Manifest.permission.READ_SMS)) {
            } else {
                ActivityCompat.requestPermissions(context, arrayOf(Manifest.permission.READ_SMS),
                        REQUEST_PERMISSION_KEY)
            }
        }
        refreshSMS()
        list.adapter = MessagesAdapter(context, mutableListOf<String?>().also { it += listOfComponents }, listOfContents,
                listOfIcons)
    }
}